OLD: Checks for splitModels()

grewered opened the pull request:
Do you want error messages when qMatrix and person.groups are no data frames or just warnings like it is now? 
if you use assert it will stop the function and throw an error, and you/I could delete the part that is now commented out. 

Should I add checks for the lists 'add' and 'cross'? they have similar "checks" in line 219f, that maybe I could skip some of them, when you are ok with error messages. if not, then I'd just leave it as it is.

grewered committed:
added checks for logical arguments

grewered committed:
qmatrix, person.groups: used checkmate checks

grewered committed:
added checks for chr arguments

grewered committed:
added checks for numeric and integer arguments

weirichs commented on R/splitModels.R:
```diff
@@ -1,5 +1,16 @@
-    if(!is.null(qMatrix))       {qMatrix      <- checkQmatrixConsistency (qMatrix)}
+  # checks
+  lapply(c(all.persons, full.model.names, include.var.name, env, verbose), checkmate::assert_logical, len = 1)
+  lapply(c(split, model.name.elements), checkmate::assert_character, null.ok = TRUE)
```
I would suggest a stricter test here, because only four values are allowed for `model.name.elements`. So maybe
```
checkmate::assert_character(split, null.ok = TRUE, max.len = 2, unique=TRUE)
checkmate::assert_character(model.name.elements, null.ok = FALSE, max.len = 4, unique=TRUE)
```

weirichs commented on R/splitModels.R:
```diff
@@ -1,5 +1,16 @@
 splitModels <- function ( qMatrix = NULL , person.groups = NULL , split = c ( "qMatrix" , "person.groups" ) , add = NULL , cross = NULL , all.persons = TRUE , all.persons.lab = "all" , person.split.depth = 0:length(person.groups[,-1,drop=FALSE]), full.model.names = TRUE , model.name.elements = c ( "dim" , "group" , "cross" ) , include.var.name = FALSE , env = FALSE , nCores=NULL , mcPackage = c(" ...
-    if(!is.null(qMatrix))       {qMatrix      <- checkQmatrixConsistency (qMatrix)}
```
I would leave this line here because the function already performs various checks for the Q matrix.

weirichs commented on R/splitModels.R:
```diff
@@ -1,5 +1,16 @@
+  checkmate::assert_numeric(GBcore, len = 1, null.ok = TRUE)
+
+  assert_data_frame(qMatrix, min.rows = 1, min.cols = 1, null.ok = TRUE)
+  assert_data_frame(person.groups, min.rows = 1, min.cols = 1, null.ok = TRUE)
```
These two lines are not necessary if `if(!is.null(person.groups)) {person.groups<- checkPersonGroupsConsistency (person.groups)}` remains in the function.

weirichs commented on R/splitModels.R:
```diff
@@ -9,52 +20,47 @@ splitModels <- function ( qMatrix = NULL , person.groups = NULL , split = c ( "q
+#		if ( !is.null ( person.groups ) & !is.data.frame ( person.groups ) ) {
+#				person.groups <- NULL
+#				warning ( paste0 ( "splitModels: person.groups is not a data.frame and will be ignored." ) , call. = FALSE )
+#		}
```
These commented out lines can be deleted.

weirichs commented on R/splitModels.R:
```diff
@@ -9,52 +20,47 @@ splitModels <- function ( qMatrix = NULL , person.groups = NULL , split = c ( "q
+#						warning ( "splitModels: check person.groups" , call. = FALSE )
+#						person.groups <- NULL
+#				}
+#		}
```
These commented out lines can be deleted.

weirichs commented on R/splitModels.R:
```diff
@@ -9,52 +20,47 @@ splitModels <- function ( qMatrix = NULL , person.groups = NULL , split = c ( "q
-								qMatrix <- qMatrix[,colnames(qMatrix)[!colnames(qMatrix) %in% names (not01)[not01]],drop=FALSE]
-						}
+		if ( test_data_frame(qMatrix, min.cols = 2) ) {
+		  not01 <- ! sapply ( qMatrix[,-1,drop=FALSE] , function ( x ) all ( x %in% c(0,1) ) )
```
I think these checks are already performed in `checkQmatrixConsistency()`. They seem a little redundant here.

weirichs commented on R/splitModels.R:
```diff
@@ -9,52 +20,47 @@ splitModels <- function ( qMatrix = NULL , person.groups = NULL , split = c ( "q
-				}
+		if ( test_data_frame(qMatrix, ncols = 1) ) {
+		  warning ( "splitModels: qMatrix contains just one column; this is treated as item names" , call. = FALSE )
+			qMatrix$dim <- 1
```
In general: Users may specify a Q matrix in `splitModels()` or `defineModel()` (or both). Both function are exported to the NAMESPACE. So it would be more efficient to capture all necessary checks in the `checkQmatrixConsistency()` function which will be called by `splitModels()` and `defineModel()`.

weirichs reviewed (commented):
Thanks, I have made some comments at various points in the script.

grewered committed:
changed checks for split, model.name.elements

grewered committed:
reordered syntax

grewered committed:
deleted unnecessary code

grewered committed:
qMatrix: moved checks from splitModels() to checkQmatrixConsistency()

grewered committed:
person.groups: moved checks from splitModels() to checkPersonGroupsConsistency()

grewered commented:
so I moved the remaining two tests to the checkXY functions and deleted the rest. If you say that's ok for now, I would deleted the commented out code in splitModels(), but I wasn't sure, whether this is what you meant, or wether these checks are already performed in the checkXY functions.

grewered commented on R/splitModels.R:
```diff
@@ -487,5 +461,19 @@ checkPersonGroupsConsistency <- function (d) {
     ### gruppierungsvariablen duerfen nicht konstant sein und keine fehlenden Werte haben
                      if(length(unique(d[,x])) == 1) {stop(paste0("Column '",x,"' of 'person.groups' is constant."))}
                      if(any(is.na(d[,x]))) {stop(paste0("Column '",x,"' of 'person.groups' has missing values."))}})
+             # wenn nur eine Spalte wird diese als IDs angenommen
```
here: checkPersonGroupsConsistency() in splitModels.R

grewered commented on R/eatModel.r:
```diff
@@ -1700,6 +1700,19 @@ checkQmatrixConsistency <-  function(qmat) {
                 cat("    "); cat(paste(qmat[weg,1],collapse=", ")); cat("\n")
                 qmat  <- qmat[-weg,]
              }
+             # wenn nur eine Spalte wird diese als IDs angenommen
```
here: checkQmatrixConsistency() in eatModel.R

weirichs commented:
> so I moved the remaining two tests to the checkXY functions and deleted the rest. If you say that's ok for now, I would deleted the commented out code in splitModels(), but I wasn't sure, whether this is what you meant, or wether these checks are already performed in the checkXY functions.

I think that should work. However, the function was originally written by Martin Hecht, so I'm not completely sure about possible consequences of the changes. Unfortunately, I don't understand all the passages in the function, so my approach in the past has always been to leave the code untouched. I have never done any damage this way. However, the code is bulky and sometimes quite redundant. Maybe you could implement your changes and then check if the examples on the help page of splitModels still work in the same way? Thanks!

weirichs reviewed (approved):
Thanks LGTM!
